cryptomite/circulant.py

```diff
@@ -0,0 +1,129 @@
+"""
+Circulant is a seeded extractor that takes an input of
+n - 1 bits and a seed of n bits, where n is prime, to
+produce some error-perfect random bits.
+"""
+from __future__ import annotations
+
+from math import floor, log2
+from typing import cast
+
+from cryptomite.utils import BitsT, closest_prime, conv, log_2
+
+__all__ = ['Circulant']
+
+
+class Circulant:
+    """ Circulant extractor based on [Foreman23]_. """
+    def __init__(self, n: int, m: int):
+        """Create a Circulant Extractor.
+
+        Parameters
+        ----------
+        n : int
+            The length of the (weak) seed bits.
+            *** This should be prime. ***
+        m : int
+            The length of the output bits.
+        """
+        self.n, self.m = n, m
+
+    def extract(self, input1: BitsT, input2: BitsT) -> BitsT:
+        """ Extract randomness.
+
+        Parameters
+        ----------
+        input1 : list of bits
+            The first list of bits, the 'input'.
+        input2 : list of bits
+            The second list of bits, the '(weak) seed'.
+
+        Returns
+        -------
+        list of bits
+            The extracted output.
+        """
+        n, m = self.n, self.m
+        assert len(input1) + 1 == len(input2) == n
+        assert n >= m
+        l = log_2(2 * n - 2)
+        L = 1 << l
+        input1.append(0)
```

Comment:
We shouldn't modify input1 without warning the user. Do something like 

`input1 = input1 + [0]`

instead, which creates a new list